### ISSUE TITLE
ActionView::Template::Error (Sort exceeded memory limit of 33554432 bytes, but did not opt in to external sorting.

### ISSUE THREAD

**Issue by Hettomei** [igrigorik/vimgolf#307]
Using `heroku logs -t | grep -i error`

We can see a lot of 

```
2021-01-29T21:33:28.922622+00:00 app[web.1]: ActionView::Template::Error (Sort exceeded memory limit of 33554432 bytes, but did not opt in to external sorting. (292)):
2021-01-29T21:33:29.011488+00:00 app[web.1]: I, [2021-01-29T21:33:29.011393 #9]  INFO -- : Completed 500 Internal Server Error in 819ms
```

Issue linked to #306

**Comment by Hettomei** [comment]
@filbranden do you think it can be because of https://github.com/igrigorik/vimgolf/commit/0828afaf1b858fc9cfa9e57106abecf1fd172e51 ?

**Comment by filbranden** [comment]
That code path should only trigger if you query a page with `?ranking=1`, so I'd imagine it's not due to that commit and you might be able to double check for that by looking at the access log and seeing whether queries including that parameter have been issued (assuming that's logged somewhere.)

Having said that, the main reason why that commit didn't work is the same as this one, the "Sort exceeded memory limit". While that one will consistently run into this issue for any non-trivial user, I guess as users post more entries in more challenges, the sort limits will probably get reached in those cases too...

I also pushed 8a1163235d4631300ec5fb32acd33c01c616cc56 which should solve this issue, as it addresses the "but did not opt in to external sorting" part, at least that's what Stack Overflow tells me, and as far as I can tell all MongoDB queries go through that code... But for some reason that doesn't really work...

One thing I would like to do is upgrade the Ruby gems to newer versions, I actually have a stack of commits locally to do that, maybe newer versions of the gem will have a fix for the problem? They build upon #301 and that one has been pending review for a while...

I just read last week that @igrigorik got a new job (Congrats!) so I guess he's been pretty busy and will probably be busy ramping up for a while... I'd be happy to try to push these PRs through. BTW, I'd be glad to help as a maintainer if you'd take me (both on GitHub, on Heroku), feel free to contact me by e-mail if you'd consider taking me into the team.

**Comment by Hettomei** [comment]
Thanks for the insight

I extracted 8 min of logs, here are the path and the errors 
[list_url.txt](https://github.com/igrigorik/vimgolf/files/5896749/list_url.txt)

**Comment by filbranden** [comment]
Yeah, I see it's happening consistently, not only for some user pages but also for some challenges which have > ~1,000 entries...

Maybe try to sync the Heroku version to just before commit 8a1163235d4631300ec5fb32acd33c01c616cc56 and see if that makes a difference? I mean, in theory that should *prevent* this issue, but I wouldn't put past it to actually make it worse...

**Comment by wjfjfm** [comment]
So, is there a command line API to get the nearest answer of recent upload? I'm suffering a lot of recent "We're sorry, but something went wrong."

**Comment by filbranden** [comment]
Also per my notes on 8a1163235d4631300ec5fb32acd33c01c616cc56, it seems the limit used to be 100MiB (error message `Sort exceeded memory limit of 104857600 bytes, but did not opt in to external sorting.`) but now you're seeing one for 32MiB, so I'm wondering if something changed on the MongoDB hosting side that reduced the memory allowance for queries and sorting?

(Also suspecting the MongoDB hosting might have something to do with the external disk sorting not working as expected...)

**Comment by Hettomei** [comment]
> So, is there a command line API to get the nearest answer of recent upload? I'm suffering a lot of recent "We're sorry, but something went wrong." [cry][cry]

No. Making an api is a topic that no one spent time to create.


@filbranden nice spot. And yes it used yo work with your code. Maybe we have to look at recent heroku change, @igrigorik have you changed something in the dyno config ?

**Comment by Hettomei** [comment]
@filbranden 
I tried to revert your 4 commits and deploy.

```
77ca84f 2021-02-19 22:59:07 +0100 Timothée Gauthier (HEAD, heroku/master, heroku/HEAD, try-fix-heroku) Revert "Allow Mongo Aggregation to use disk for temporary space. (#297)"
6390ec5 2021-02-19 22:59:06 +0100 Timothée Gauthier Revert "Display challenges in user page in chronological order (#296)"
c68451c 2021-02-19 22:59:04 +0100 Timothée Gauthier Revert "Display all challenge solutions per user (#294)"
a368483 2021-02-19 22:58:52 +0100 Timothée Gauthier Revert "Include position in ranking for each challenge in user page (#293)"
```

Deploy was ok but we have the same error on heroku.

So my new ideas based on https://developerslogblog.wordpress.com/2019/11/27/mongodb-sort-how-to-fix-maximum-ram-exceeded-error/ : maybe before it was ok to not have any index. Now it fails just because database is bigger.

I m not good at mongodb, what do you think ?

**Comment by Hettomei** [comment]
In fact, in case I failed to properly revert, I deployed commit `68195d5 2020-06-24 14:31:41 -0700 dependabot[bot] Bump rack from 2.0.8 to 2.2.3 (#290)` and server return the same error.

**Comment by filbranden** [comment]
I'm not great at MongoDB earlier...

Yeah I think maybe an index might help, but I don't have much idea of *which* index. I might try to play with it another while...

I think the default limit for external sorts (from previous messages) is 100MB, so I *suspect* what might have happened here is whoever is hosting MongoDB (Heroku? or a third party?) might have tuned it to allow a smaller in-memory amount. They might have blocked disk access too...

I think there are some commands you can use to see the MongoDB configuration (I found that while googling for that), if you have access to some kind of console, perhaps you can find some of that out?

**Comment by Hettomei** [comment]
Based on https://developerslogblog.wordpress.com/2019/11/27/mongodb-sort-how-to-fix-maximum-ram-exceeded-error/ I tried the following : 

```
 app/models/challenge.rb

  index "entries.score" => 1
  index "entries.created_at" => 1
  index({ "entries.score": 1, "entries.created_at": 1 })
```

deploy + 

```
heroku run "rake db:mongoid:create_indexes"
MONGOID: Created indexes on Challenge:
MONGOID: Index: {:user_id=>1}, Options: {:background=>true}
MONGOID: Index: {:"entries.user_id"=>1}, Options: {}
MONGOID: Index: {:"entries.score"=>1}, Options: {}
MONGOID: Index: {:"entries.created_at"=>1}, Options: {}
MONGOID: Index: {:"entries.score"=>1, :"entries.created_at"=>1}, Options: {}
```


```
 heroku run "rails c"
Loading production environment (Rails 5.0.3)
irb(main):001:0> Challenge.collection.indexes.to_a
=>
[{"v"=>2, "key"=>{"_id"=>1}, "name"=>"_id_"}, 
{"v"=>2, "key"=>{"user_id"=>1}, "name"=>"user_id_1", "background"=>true}, 
{"v"=>2, "key"=>{"entries.user_id"=>1}, "name"=>"entries.user_id_1"}, 
{"v"=>2, "key"=>{"entries.score"=>1}, "name"=>"entries.score_1"}, 
{"v"=>2, "key"=>{"entries.created_at"=>1}, "name"=>"entries.created_at_1"}, 
{"v"=>2, "key"=>{"entries.score"=>1, "entries.created_at"=>1}, "name"=>"entries.score_1_entries.created_at_1"}]
```

still the same error

**Comment by Hettomei** [comment]
It looks like challenge with roughly less than 500 entries is ok

+ works : Sorting paragraphs - 456 entries https://vimgolf.herokuapp.com/challenges/4d1b3d57c58eaa2a8a000510 
+ fail : Interweave two blocks of text - 530 entries  https://vimgolf.herokuapp.com/challenges/4dcd7b572c8e510001000005

**Comment by Hettomei** [comment]
@igrigorik 
Can you forward us any mongodb logs ?

**Comment by igrigorik** [comment]
Following the exception trail.. From Rails:

```
2021-02-28T23:12:58.048737+00:00 app[web.1]: I, [2021-02-28T23:12:58.048608 #9]  INFO -- : Completed 500 Internal Server Error in 151ms
2021-02-28T23:12:58.052773+00:00 app[web.1]:   
2021-02-28T23:12:58.052945+00:00 app[web.1]: ActionView::Template::Error (Sort exceeded memory limit of 33554432 bytes, but did not opt in to external sorting. (292)):
2021-02-28T23:12:58.053171+00:00 app[web.1]:     11:     <p>Check out <a href="/about#resources"> these helpful resources</a> to improve your Vim skills... <em>Game on.</em></p>
2021-02-28T23:12:58.053172+00:00 app[web.1]:     12: 
2021-02-28T23:12:58.053173+00:00 app[web.1]:     13: <% if !current_user %>
2021-02-28T23:12:58.053174+00:00 app[web.1]:     14:     <div class="note clearfix">Unlock <b><%= @submissions.count_remaining %></b> remaining solutions by <b><%= link_to "signing in", "/auth/twitter?x_auth_access_type=read&use_authorize=true" %></b> and submitting your own entry</div>
2021-02-28T23:12:58.053175+00:00 app[web.1]:     15:   <% elsif @submissions.count_remaining > 0 %>
2021-02-28T23:12:58.053179+00:00 app[web.1]:     16:     <div class="note clearfix">Unlock <b><%= @submissions.count_remaining %></b> remaining solutions by submitting a higher ranked entry</div>
2021-02-28T23:12:58.053182+00:00 app[web.1]:     17: <% end %>
2021-02-28T23:12:58.053329+00:00 app[web.1]:   
2021-02-28T23:12:58.053380+00:00 app[web.1]: app/repositories/repository_challenge.rb:216:in `first'
2021-02-28T23:12:58.053381+00:00 app[web.1]: app/repositories/repository_challenge.rb:216:in `worst_score'
2021-02-28T23:12:58.053381+00:00 app[web.1]: app/services/submissions.rb:65:in `worst_score'
2021-02-28T23:12:58.053382+00:00 app[web.1]: app/services/submissions.rb:46:in `visible_score'
2021-02-28T23:12:58.053382+00:00 app[web.1]: app/services/submissions.rb:99:in `count_remaining'
```

Except, the function we call does pass allow_disk_use: 

https://github.com/igrigorik/vimgolf/blob/0828afaf1b858fc9cfa9e57106abecf1fd172e51/app/repositories/repository_challenge.rb#L26-L28

Did the driver change and the option is a no op? I'm having a hard time with the docs to even find it. grr!

**Comment by Hettomei** [comment]
@igrigorik 
Thanks for rails log.

But it is possible that mongodb's log are more precise if they exist. 
Do you have access to it ?

**Comment by igrigorik** [comment]
For the life of me, can't figure out where to get access to them in MongoDB Atlas.. Do you happen to know?

**Comment by Hettomei** [comment]
I have no idea.

And I tried to connect to mongo atlas with a mongo client using vimgolf's credentials, but I think the client must be 'declared' or in the same network so I can't connect from home.

**Comment by filbranden** [comment]
This might be a limitation of MongoDB Atlas Free Tier...

https://docs.atlas.mongodb.com/reference/free-shared-limitations/#operational-limitations

"Atlas Free Tier and shared clusters do not support the `allowDiskUse` option for the aggregation command or its helper method."

**Comment by igrigorik** [comment]
> This might be a limitation of MongoDB Atlas Free Tier (M0) or shared clusters (M2, M5)... You might need a dedicated M10+ instance for that feature.

Well, doh! We're on M5 — grr. 

I guess the real puzzle is what changed recently to start causing this.. Unless Atlas recently changed their policy on this?

**Comment by filbranden** [comment]
Looks like M0/M2/M5 instances were upgraded to MongoDB 4.4 in January:

https://docs.atlas.mongodb.com/release-notes/atlas/#26-january-2021-release

They don't mention tweaks to the limits, but it's possible that this change made a difference here.

Allowing disk use for sort hasn't been enabled for a long time, but maybe the behavior of older version of MongoDB would always allow it or use whatever resources are needed to fulfill the query...

I managed to reproduce this locally, on a test challenge with 37,251 entries (I was doing some benchmarking a while ago) and reverting the `allow_disk_use` commit in my local repository. I see it fail on the query for `best_player_score()`. I looked at whether it was possible to add more indices to alleviate the issue, but since that query creates a new calculated field `min_score` and sorts on it, I'm not sure how to address that...

Long term, I think #298 is the path forward. I have looked at that in the past and I'm happy to try to put some more time in trying to wrap that up. There's, of course, a migration to happen for that. But I'd be happy to look into all of that. What do you say?

**Comment by igrigorik** [comment]
A short-term fix is to upgrade or move to a service that supports allow_disk_use. Besides Atlas, do you guys have any recommendations for hosted Mongo?

**Comment by Hettomei** [comment]
@filbranden 

A silly temporary solution is to massively reduce the database size for a short period : 

Copy the database.

Remove half of all the answers. (Or something smarter)

The website works again. (Maybe)

Then you agregate the new answers with the old database when everything is fixed :)

**Comment by Hettomei** [comment]
@filbranden 

Maybe we can make the site working again by removing all the count.

Let say, we start a branch "degraded-experience" where : 

We push this branch on heroku (and keep master clean so you can work on postgre)

##Home page :
We display a banner "Site temporary on low functionnality. We need maintener ..."

## Profile page : 
Only the title of the challenge already done 

(I discover that it works on my profile ? if it s ok for everyone I change nothing)

## Challenge page
No leaderboard

Only the "previous solution"


-----------

When the moving from mongodb to any new solution is working, we delete this branch ang go back to master. 

So people can still play. I think it is better to have a degraded experience than nothing.

I can do this work.

What do you think ?

**Comment by filbranden** [comment]
@Hettomei Yes it's a good idea.

I explored limiting the entries with an early `{ '$limit': 400 }` before any sorting but it didn't work... Perhaps just dropping all leaderboards and scores for now would at least allow people to access the challenges themselves...

**Comment by Hettomei** [comment]
@filbranden I read somewhere that

"if you use      { "$unwind": "$entries" } , then you have a bad datastructure"
(source lost, was on stackoverflow with some body trying to agregate without allowDiskUse)

And it looks like we can't apply an index in the embeded document Challenge.entries.

My next try will be to make "Entry" a document, and adding a "challenge_id" (which, of course, will point to the current Document)

wait & see.

**Comment by Hettomei** [comment]
Looks like having all "entries" at root level seems to fix the problem. See https://github.com/igrigorik/vimgolf/pull/315

It gives me hope

**Comment by filbranden** [comment]
Fixed after migration to Postgres (#298).
